README.md
# Online_Auction_Portal
Online Auction Portal in C using Socket Programming


DATABASE/history.txt
son	1	2	Fri.May..9.09:40:46.2014
ha	1	20	Fri.May..9.09:52:46.2014
ha	sony	20	Fri.May..9.09:58:15.2014
son	Car	33000	Mon.May.12.18:50:11.2014
son	Car	33000	Tue.May.13.15:20:09.2014
son	ElectricDoor	350	Wed.May.14.09:41:40.2014
ha	ElectricDoor	3380	Wed.May.14.09:48:12.2014
ha	Electric	650	Tue.May.20.23:10:04.2014


CLIENT/client.c
#include<sys/types.h>
#include<sys/socket.h>
#include<stdio.h>
#include<stdlib.h>
#include<netinet/in.h>
#include<arpa/inet.h>
#include<unistd.h>
#include<sys/ioctl.h>
#include<string.h>
#include<netdb.h>
#include <fcntl.h>
#include <pthread.h>

#define SIG_FAIL      -6
#define SIG_SUCCESS   -5
#define SIG_LOWBID    -4
#define SIG_NEM       -3
#define SIG_ALI       -2
#define SIG_EXCEPTION -1
#define CMD_REGISTER   0
#define CMD_SIGNIN     1
#define CMD_JOIN       2
#define CMD_HISTORY    3
#define CMD_BID        4
#define CMD_SIGNOUT    5
#define CMD_END        6
#define CMD_WINNER     7
#define CMD_DEPOSIT    8
#define CMD_REJECT     9
#define STDIN          0

void menu();
char* getServerIP();
int menu1_option1();    //Sign in
int menu1_option2();    //Sign up
int menu2_option1();    //Get History
int menu2_option2();    //Join auction
int menu2_option3();    //Deposit money
int menu2_option4();    //Sign out

int sockfd; //socket of server file descriptor

typedef struct{
    char name[40];
    char password[50];
    int  balance;
}User;

User user;
int  price;

int main(){
    int     retval;
    int     option;
    int     menu_type=1;
    struct sockaddr_in address;
    //Create socket to connect to server
    sockfd = socket(AF_INET,SOCK_STREAM,0);
    if(sockfd == 0){
        perror("socket() error\n");
    }
    //Server ip address
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(getServerIP());
    address.sin_port = htons(8888);

    retval = connect(sockfd,(struct sockaddr*)&address,sizeof(address));
    //Connection error
    if(retval == -1){
        perror("connect() error!\n");
        close(sockfd);
        return -1;
    }
    do{
        menu:
        menu(menu_type);
        scanf("%d",&option);
        getchar();
        switch(option){
            case 1:
                if(menu_type==1){
                    if(menu1_option1())
                        menu_type = 2;
                } else {
                    menu2_option1();
                }
                break;
            case 2:
                if(menu_type==1){
                    if(menu1_option2())
                        menu_type = 2;
                } else {
                    menu2_option2();
                    menu_type = 2;
                    goto menu;
                }
                break;
            case 3:
                if(menu_type==1){
                    close(sockfd);
                    return 0;
                } else {
                    menu2_option3();
                }
                break;
            case 4:
                if(menu_type==2){
                    if(menu2_option4()){
                        menu_type = 1;
                    }
                }
                else
                    printf("Please enter a valid option\n");
                break;
            default:
                printf("Please enter a valid option\n");
        }
    } while(1);
    return 0;
}

void menu(int type){
    if(type==1){
        printf("\n*** AUCTION CLIENT ***\n\n");
        printf("   1. Sign In\n");
        printf("   2. Sign Up\n");
        printf("   3. Exit\n\n");
        printf("Option: ");
    } else if(type == 2){
        printf("\n************************************\n");
        printf("\nHello! %s\n",user.name);
        printf("Your balance is %d RS\n\n", user.balance);
        printf("   1. Join Auction\n");
        printf("   2. History\n");
        printf("   3. Deposit money\n");
        printf("   4. Log Out\n\n");
        printf("Option: ");
    }
}

int menu1_option1(){    //Sign in
    int command;        //Variable to receive command from client
    User temp;
    char line[100];
    printf("Username:");    scanf("%s",temp.name);
    printf("Password:");    scanf("%s",temp.password);
    command = CMD_SIGNIN;
    write(sockfd,&command,sizeof(int));
    write(sockfd,&temp,sizeof(User));
    read(sockfd,&command,sizeof(int));
    if(command == SIG_EXCEPTION){
        printf("Incorrect username or password.\n");
        return 0;
    } else if(command == SIG_ALI){
        printf("Username %s already logged in!\n",temp.name);
        return 0;
    } else {
        //receive balance of user
        read(sockfd,&user.balance,sizeof(int));
        printf("\n************************************\n");
        printf("\nGoods for auction at the moment:\n");
        read(sockfd,line,sizeof(char)*100);
        printf("%s\n",line);
        strcpy(user.name,temp.name);
        strcpy(user.password,temp.password);
    }
    return 1;
}

int menu1_option2(){    //Sign up
    int command;        //Variable to receive command from client
    User temp;
    char password[50];
    char line[100];
    while(1){
        printf("New Username:");    scanf("%s",temp.name);
        if(strcmp(temp.name,"")==0){
            printf("Username must not be empty\n");
        } else {
            break;
        }
    }
    while(1){
        printf("Password:");        scanf("%s",temp.password);
        printf("Password Confirmation:");   scanf("%s",password);
        if(strcmp(temp.password,password)!=0){
            printf("Password and confirmation do not match.\n");
        } else {
            break;
        }
    }
    command = CMD_REGISTER;
    write(sockfd,&command,sizeof(int));
    write(sockfd,&temp,sizeof(User));
    read(sockfd,&command,sizeof(int));
    if(command != CMD_REGISTER){
        printf("Username already existed.\n");
        return 0;
    } else {
        printf("Register successfully!\n");
        printf("\nHello! %s\n",temp.name);
        printf("Goods for auction at the moment:\n");
        //receive balance of user
        read(sockfd,&user.balance,sizeof(int));
        read(sockfd,line,sizeof(char)*100);
        printf("%s",line);
        strcpy(user.name,temp.name);
        strcpy(user.password,temp.password);
    }
    return 1;
}

int menu2_option2(){    //Get History
    char line[1024];
    int command;
    command = CMD_HISTORY;
    write(sockfd,&command,sizeof(int));
    read(sockfd,line,sizeof(char)*1024);
    printf("%s\n",line);
    return 1;
}

int menu2_option1(){    //Join auction
    int command;
    char choice;
    char line[100];
    char buffer[200] = "";  //buffer for stdin
    int  retval;
    int  byte_count;
    int  bid_money;
    int  n;
    int warning = 0;
    fd_set allfds;
    fd_set readfds;
    command = CMD_JOIN;
    printf("You can't quit once you join the auction. Do you want to join?\n");
    printf("Join(1)/Cancel(any key)?:");    scanf("%c",&choice);
    if(choice == '1'){
        //Initiate read file descriptor
        FD_ZERO(&readfds);
        FD_ZERO(&allfds);
        FD_SET(sockfd,&allfds);
        FD_SET(STDIN,&allfds);
        //Send command, receive, price and goods infor
        write(sockfd,&command,sizeof(int));
        printf("Joining...\n");
        read(sockfd,&command,sizeof(int));
        if (command == CMD_REJECT)
        {
            printf("\n No auction is being held at the moment.\n\n");
            return 1;
        }
        read(sockfd,&price,sizeof(int));
        read(sockfd,line,sizeof(line));
        printf("%s",line);
        //read(sockfd,line,sizeof(line));
        //printf("%s",line);
        strcpy(line,"");
        printf("\n\n");
        printf("\nBid: "); fflush(stdout);
        while(1){
            readfds = allfds;
            retval = select(FD_SETSIZE,&readfds,NULL,NULL,NULL);
            if(retval == -1){
                perror("select() error\n");
                return -1;
            }
            //Activity triggered by server
            if(FD_ISSET(sockfd,&readfds)){
                ioctl(sockfd,FIONREAD,&byte_count);
                if(byte_count==0){
                    printf("Connection lost!\n");
                    exit(1);
                } else {
                    read(sockfd,line,sizeof(char)*100);
                    if(strlen(line) > 1){
                        printf("%s",line);
                    } else{
                        command = atoi(line);
                        if(command == CMD_END){
                            printf("\rEnd of auction !\n");
                            break;
                        } else if(command == CMD_WINNER){
                            read(sockfd,&user.balance,sizeof(int));
                            printf("\rEnd of auction !\n");
                            break;
                        }
                    }
                    printf("\nBid: "); fflush(stdout);


                }
              //Activity triggered by stdin
            } else if(FD_ISSET(STDIN,&readfds)){
                if ((n = read(STDIN, buffer, sizeof(char)*200)) != 0) {
                    bid_money = atoi(buffer);
                    if(bid_money==0){
                        printf("\rPlease enter a valid number\n");
                        warning += 1;
                        if (warning == 5)
                        {
                            printf("\r You have tried invalid input for 5 times, and forced to quit.\n");
                            exit(0);
                        }
                    } else {
                        command = CMD_BID;
                        write(sockfd,&command,sizeof(int));
                        write(sockfd,&bid_money,sizeof(int));
                        read(sockfd,&command,sizeof(int));
                        if(command == SIG_NEM){
                            printf("Your bidding money exceeds your balance\n");
                            warning += 1;
                            if (warning == 5)
                            {
                                printf("\r You have tried invalid input for 5 times, and forced to quit.\n");
                                exit(0);
                            }
                        } else if(command == SIG_LOWBID){
                            printf("Your bidding money is lower than current price + minimum increment\n");
                            warning += 1;
                            if (warning == 5)
                            {
                                printf("\r You have tried invalid input for 5 times, and forced to quit.\n");
                                exit(0);
                            }
                        }
                        else warning = 0;
                    }
                    printf("\nBid: "); fflush(stdout);
                }
            }

        }
    }
    return 1;
}

int menu2_option3(){    //Deposit money
    char serial[100];
    int command;
    int money;
    printf("Enter serial ID:");
    scanf("%s",serial);
    command = CMD_DEPOSIT;
    write(sockfd,&command,sizeof(int));
    write(sockfd,serial,sizeof(char)*100);
    read(sockfd,&command,sizeof(int));
    if(command == SIG_SUCCESS){
        read(sockfd,&money,sizeof(int));
        printf("%dRS has been deposited\n",money);
        user.balance += money;
    } else if(command == SIG_FAIL){
        printf("Your serial ID is not correct\n");
    }
    return 1;
}

int menu2_option4(){    //Sign out
    int command;
    command = CMD_SIGNOUT;
    printf("%s has logged out!\n",user.name);
    strcpy(user.name,"");
    strcpy(user.password,"");
    write(sockfd,&command,sizeof(int));
    return 1;
}

char* getServerIP()
{
    FILE *fi;
    char IP[17];

    if((fi = fopen("../DATABASE/server_addr.txt","r"))==NULL)  {
        printf("Error opening serverIP file!\n");
        return NULL;
    }

    fscanf(fi,"%s",IP);
    fclose(fi);
    return strdup(IP);

}


ADMIN_SERVER/server_utilities.h
#include<sys/socket.h>
#include<stdio.h>
#include<stdlib.h>
#include<netinet/in.h>
#include<sys/types.h>
#include<sys/time.h>
#include<time.h>
#include<sys/ioctl.h>
#include<unistd.h>
#include<string.h>
#include<ctype.h>
#include <pthread.h>

#define SIG_FAIL      -6
#define SIG_SUCCESS   -5
#define SIG_LOWBID    -4    //Bidding money lower than current price + min increment
#define SIG_NEM       -3    //Not enough money
#define SIG_ALI       -2    //Already logged in
#define SIG_EXCEPTION -1
#define CMD_REGISTER   0
#define CMD_SIGNIN     1
#define CMD_JOIN       2
#define CMD_HISTORY    3
#define CMD_BID        4
#define CMD_SIGNOUT    5
#define CMD_END        6
#define CMD_WINNER     7
#define CMD_DEPOSIT    8
#define CMD_REJECT     9
#define STT_OFFLINE    0
#define STT_ONLINE     1
#define STT_BIDDING    2

typedef struct {
    char name[50];
    int  init_price;
    int  cur_price;
    int  min_incr;
}Goods;

typedef struct{
    char name[40];
    char password[50];
    int  balance;
    int  status;        //0:offline, 1:online, 2:bidding
}User;

typedef struct{
    int client;
    int bid_money;
}LogEntry;

// Administrating
void importDB();
void printGoodList();
void exportDB();
void enterGoods();
void editGoods();
void deleteGoods();
int chooseAuctionItem();
int editMenu();
void viewHistory();

// Utilities
int registerUser(User* user);
User* getUserByName(char *name);
char* getGoodsinfo();
char* getHistory(char* user_name);
void broadcast(char* str);
void broadcastEnd(int winner);
int authenticate(User *user);
int isLoggedIn(char* name);
int updateUser(User user);
int writeLog(char *name, char* goods, int price,char* datetime);
char* replace(char* str,char old, char new);
int deposit(User user,char* serial);
int MenuChoice();
void* menuThread(void* threadid);


ADMIN_SERVER/server_utilities.c
#include "server_utilities.h"

extern Goods goods;            //Goods to auction
extern Goods goodslist[100];
extern int good_count;     // Total number of goods in DB
extern int auction_state;      //Default = 0, after first person join = 1,
                        //during 1st 20s of bid = 2,during 2nd 20s of bid = 3, during 3rd 20s of bid = 4
extern int isCount;
extern User *users;
extern LogEntry logging[500];

int registerUser(User* user){
    FILE *file;
    if(getUserByName(user->name) != NULL){
        return 0;
    }
    if((file = fopen("../DATABASE/bidders.txt","a"))==NULL)  {
        printf("Error opening file!\n");
        return -1;
    }
    user->balance = 500000;
    fprintf(file,"%s\t%s\t%d\n",user->name,user->password,user->balance);
    fclose(file);
    return 1;
}

int updateUser(User user){
    FILE *file;
    User temp[100];
    int n = 0;
    int i;
    if((file = fopen("../DATABASE/bidders.txt","r"))==NULL)  {
        printf("Error opening file!\n");
        return 0;
    }
    while(fscanf(file,"%s %s %d",temp[n].name,temp[n].password,&temp[n].balance)!= EOF){
        if(strcmp(user.name,temp[n].name)==0){
            temp[n].balance = user.balance;
        }
        n++;
    }
    fclose(file);
    if((file = fopen("../DATABASE/bidders.txt","w"))==NULL)  {
        printf("Error opening file!\n");
        return 0;
    }
    for(i = 0; i<n; i++){
        fprintf(file,"%s\t%s\t%d\n",temp[i].name,temp[i].password,temp[i].balance);
    }
    fclose(file);
    return 1;
}

int authenticate(User *user){
    User* temp;
    temp = getUserByName(user->name);
    if(temp==NULL)  return 0;
    if(strcmp(user->password,temp->password)==0){
        user->balance = temp->balance;
        return 1;
    }
    return 0;
}

void broadcastEnd(int winner){
    int i;
    int command = CMD_END;
    char line[100];
    sprintf(line,"%d",command);
    for(i = 0; i<FD_SETSIZE; i++){
        if(users[i].status == 2){
            if(i!=winner){
                write(i,line,sizeof(char)*100);
            }
            else{
                command = CMD_WINNER;
                sprintf(line,"%d",command);
                write(i,line,sizeof(char)*100);
                command = CMD_END;
                sprintf(line,"%d",command);
            }
        }
    }
}

void broadcast(char* str){
    int i;
    for(i = 0; i<FD_SETSIZE; i++){
        if(users[i].status == 2){
            write(i,str,sizeof(char)*100);
        }
    }
}

User* getUserByName(char *name){
    FILE *file;
    User* user;
    if((file = fopen("../DATABASE/bidders.txt","r"))==NULL)  {
        printf("Error opening file!\n");
        return NULL;
    }
    user = (User*)malloc(sizeof(User));
    while(fscanf(file,"%s %s %d",user->name,user->password,&user->balance)!= EOF){
        if(strcmp(user->name,name)==0){
            return user;
        }
    }
    fclose(file);
    return NULL;
}

int isLoggedIn(char* name){
    int i;
    for(i = 0; i<FD_SETSIZE; i++){
        if(strcmp(name,users[i].name)==0)
            return 1;
    }
    return 0;
}

char* getGoodsinfo(){
    char line[100];
    if (goods.cur_price == 0) sprintf(line,"\n  No auction is being held at the moment.\n");
    else sprintf(line,"  Goods: \t\t%s\n  Initial Price: \t%d RS\n  Current Price: \t%d RS\n  Step Size: \t\t%d RS\n",goods.name,goods.init_price,goods.cur_price,goods.min_incr);
    line[strlen(line)-1] = '\0';
    return  strdup(line);
}

int MenuChoice()
{
    int choice = 0;
    printf("**** AUCTION CONTROL MAIN MENU **** ");
    printf("\n\n");
    printf("1. Choose item to start auction.\n");
    printf("2. Edit item database.\n");
    printf("3. View auction history.\n");
    printf("4. Exit.\n");
    printf("\n Your choice (1-4): ");
    scanf("%d",&choice);
    while (!((choice >=1)&&(choice <=4)))
    {
        printf("\n Invalid choice. Please input a number from 1 to 4: ");
        while ( getchar() != '\n' );
        scanf("%d",&choice);
    }
    return choice;
}

void importDB()
{
    FILE *fi;
    int n = 0;
    Goods tmp;

    if((fi = fopen("../DATABASE/all_items.txt","r"))==NULL)  {
        printf("Error opening goods file!\n");
        return;
    }

    while (fscanf(fi,"%d\t%d\t",&tmp.init_price,&tmp.min_incr)>0)
        {
            fgets(tmp.name,40,fi);
            tmp.name[strlen(tmp.name)-1]='\0';
            goodslist[n].init_price = tmp.init_price;
            goodslist[n].min_incr = tmp.min_incr;
            strcpy(goodslist[n].name, tmp.name);
            n++;
        }
    fclose(fi);
    good_count = n;
}

void exportDB()
{
    FILE *fo;
    int i=0;
    if((fo = fopen("../DATABASE/all_items.txt","w"))==NULL)  {
        printf("Error opening goods file!\n");
        return;
    }

    for (i=0;i<good_count;i++)
            fprintf(fo,"%d\t%d\t%s\n",goodslist[i].init_price,goodslist[i].min_incr,goodslist[i].name);

    fclose(fo);
}


int writeLog(char *name, char* goods, int price,char* datetime){
    FILE *file;
    if((file = fopen("../DATABASE/history.txt","a+"))==NULL)  {
        printf("Error opening file!\n");
        return 0;
    }
    fprintf(file,"%s\t%s\t%d\t%s",name,goods,price,datetime);
    fclose(file);
    return 1;
}

int deposit(User user,char* serial){
    FILE *file;
    char str[100];
    int flag = 0;
    int money = 0;
    if((file = fopen("../DATABASE/serial_list.txt","r"))==NULL)  {
        printf("Error opening file!\n");
        return 0;
    }
    while(fscanf(file,"%s %d",str,&money)!= EOF){
        if(strcmp(serial,str)==0){
            flag = 1;
            break;
        }
    }
    fclose(file);
    if(!flag){
        return 0;
    }
    user.balance += money;
    updateUser(user);
    return money;
}

void printGoodList()
{
    int i=0;

    printf("\n");
    printf("|-----|------------------------------------|---------------|---------------|\n");
    printf("| No. |              Item name             |  Start price  | Min increment |\n");
    printf("|-----|------------------------------------|---------------|---------------|\n");
    if (good_count < 1)
    printf("|                        There is no item in database !                    |\n");
    for (i=0; i<good_count; i++)
        printf("| %3d | %-34s | %11d   | %11d   |\n",i+1,goodslist[i].name,goodslist[i].init_price,goodslist[i].min_incr);
    printf("|-----|------------------------------------|---------------|---------------|\n");
}

int chooseAuctionItem()
{
    int k = 0;
    printf("\n");
    if (good_count < 1)
        {
            printf(" Press add an item to the database first !\n");
            return 0;
        }

    printf("    Input No. of item to auction: (1-%d): ",good_count); scanf("%d",&k);
    while ((k < 1) || (k > good_count))
    {
        printf("Invalid item No. Please input a number from 1 to %d : ",good_count);
        scanf("%d",&k);
    }

    goods.init_price = goodslist[k-1].init_price;
    goods.min_incr = goodslist[k-1].min_incr;
    goods.cur_price = goods.init_price;
    strcpy(goods.name, goodslist[k-1].name);
    printf("Item '%s' with start price %d$ and minimum increment bid %d$ is selected !\n",goods.name,goods.init_price,goods.min_incr);
    return 1;
}

int editMenu()
{
    int choice = 0;
    printf("**** EDITTING ITEM DATABSE **** ");
    printf("\n\n");
    printf("1. Add new item.\n");
    printf("2. Edit item.\n");
    printf("3. Delete item.\n");
    printf("4. Back to main menu.\n");
    printf("\n Your choice (1-4): ");
    scanf("%d",&choice);
    while (!((choice >=1)&&(choice <=4)))
    {
        printf("\n Invalid choice. Please input a number from 1 to 4: ");
        while ( getchar() != '\n' );
        scanf("%d",&choice);
    }
    return choice;
}

void enterGoods(){
    importDB();
    printf("** Enter detail of new item **\n\n");
    while ( getchar() != '\n' );
    printf("Name of the goods: ");      gets(goodslist[good_count].name);
    printf("Initial price: ");          scanf("%d",&goodslist[good_count].init_price);
    printf("Minimum increment:");       scanf("%d",&goodslist[good_count].min_incr);
    good_count ++;
    exportDB();
    printf("\nItem added to database successfully !\n");
    printGoodList();
}

void editGoods(){
    importDB();
    int choice = 0;
    printGoodList();
    printf("\n\n");
    printf("**** Enter No. of the item to edit ****");
    printf("\n\n");
    scanf("%d",&choice);
    while (!((choice >=1)&&(choice <=good_count)))
    {
        printf("\n Invalid choice. Please input a number from 1 to %d: ",good_count);
        scanf("%d",&choice);
    }
    printf("\n");
    while ( getchar() != '\n' );
    printf("New name of the goods: ");      fgets(goodslist[choice-1].name,50,stdin);   goodslist[choice-1].name[strlen(goodslist[choice-1].name)-1]='\0';
    printf("New initial price: ");          scanf("%d",&goodslist[choice-1].init_price);
    printf("New minimum increment:");       scanf("%d",&goodslist[choice-1].min_incr);

    exportDB();
    printf("\nItem editted successfully !\n");
    printGoodList();
}

void deleteGoods(){
    importDB();
    int i = 0, choice = 0;
    printGoodList();
    printf("\n\n");
    printf("**** Enter No. of the item to delete ****");
    printf("\n\n");
    scanf("%d",&choice);
    while (!((choice >=1)&&(choice <=good_count)))
    {
        printf("\n Invalid choice. Please input a number from 1 to %d: ",good_count);
        scanf("%d",&choice);
    }
    printf("\n");
    good_count--;
    for (i = choice-1; i < good_count; i++)
    {
        goodslist[i].init_price = goodslist[i+1].init_price;
        goodslist[i].min_incr = goodslist[i+1].min_incr;
        strcpy(goodslist[i].name, goodslist[i+1].name);
    }

    exportDB();
    printf("\nItem deleted successfully !\n");
    printGoodList();
}

char* getHistory(char* user_name){
    char history[1024];
    char line[100];
    char name[40];
    char goods_name[50];
    char price[10];
    char* datetime = (char*)malloc(sizeof(char)*30);
    FILE *file;
    if((file = fopen("../DATABASE/history.txt","r"))==NULL)  {
        printf("Error opening file!\n");
        return NULL;
    }
    sprintf(history,"\rGoods                         \tPrice     \tDate&Time\n");
    while(fscanf(file,"%s %s %s %s",name,goods_name,price,datetime)!= EOF){
        if(strcmp(name,user_name)==0){
            datetime = replace(datetime,'.',' ');
            strcat(price," RS");
            sprintf(line,"%-30s\t%-10s\t%s\n",goods_name,price,datetime);
            strcat(history,line);
        }
    }
    fclose(file);
    return strdup(history);
}

char* replace(char* str,char old, char new){
    int n = strlen(str);
    int i;
    char* str1 = strdup(str);
    for(i=0; i<n; i++){
        if(str1[i]==old){
            str1[i] = new;
        }
    }
    return str1;
}


void viewHistory()
{
    FILE *fi;
    char line[100];

    if((fi = fopen("../DATABASE/history.txt","r"))==NULL)  {
        printf("Error opening auction history file!\n");
        return;
    }

    printf("\n\n");
    printf("**** Auction History ****");
    printf("\n\n");

    while (fgets(line,40,fi)>0)
    {
        printf("%s",line);
    }
    printf("\n\n");
    fclose(fi);
}

void* menuThread(void* threadid)
{   int choice = 0;

main_menu:
    choice = MenuChoice();
    switch (choice)
    {
        case 1: // Choose item, then start auction
            importDB();
            printGoodList();
            if (chooseAuctionItem() == 0) goto main_menu;
             do{
                printf("\n***Press Enter to immediately begin auction, or input 'b' to back to menu***\n");
                while ( getchar() != '\n' );
                choice = getchar();
                if (choice == 'b') goto main_menu;
             }
             while(choice!='\n');
             isCount = 1;
            break;
        case 2: // Edit item database
            edit_menu:
            choice = editMenu();
            switch (choice)
            {
                case 1: enterGoods(); goto edit_menu; break;
                case 2: editGoods(); goto edit_menu; break;
                case 3: deleteGoods(); goto edit_menu; break;
                case 4: goto main_menu;
            }
            break;
        case 3: // View auction history
            viewHistory();
            goto main_menu;
            break;
        case 4: // Exit
            exit(0); break;
    }
    return NULL;
}

